lib/kernel.js

```diff
@@ -16,25 +16,107 @@ import store from "./store";
 import WatchesStore from "./store/watches";
 import OutputStore from "./store/output";
 import HydrogenKernel from "./plugin-api/hydrogen-kernel";
+import type {
+  HydrogenKernelMiddlewareStack,
+  HydrogenKernelMiddleware
+} from "./plugin-api/hydrogen-types";
 import InputView from "./input-view";
 import KernelTransport from "./kernel-transport";
 import type { ResultsCallback } from "./kernel-transport";
 
-interface KernelMiddlewareStack {
-  interrupt: () => void;
-  shutdown: () => void;
-  restart: (onRestarted: ?Function) => void;
-  execute: (
+// Adapts middleware objects provided by plugins to an internal interface. In
+// particular, this implements fallthrough logic for when a plugin defines some
+// methods (e.g. execute) but doesn't implement others (e.g. interrupt). Note
+// that HydrogenKernelMiddleware objects are mutable: they may lose/gain methods
+// at any time, including in the middle of processing a request.
+// TODO(nikita): insert checks between plugins that ensure that messages being
+// passed around via results callbacks comply with the Jupyter messaging spec.
+class MiddlewareAdapter implements HydrogenKernelMiddlewareStack {
```

Comment:
I really really display the name `HydrogenKernelMiddlewareStack`. The first thing that `stack` brings to my mind is the concept of container. `HydrogenKernelMiddlewareStack` doesn't implement a container. It uses a stack to implement its functionality, but it isn't a container.

Reply:
I can't say I like the name either. Haven't come up with a better one, though. Any suggestions?

Reply:
I still think there is no need for `HydrogenKernelMiddlewareStack`. Hopefully, we'll get rid of it and we won't have to think of a name.